File: backend/security-oauth/src/main/java/io/featurehub/web/security/oauth/OAuth2JerseyClient.java
package io.featurehub.web.security.oauth;

import cd.connect.app.config.ConfigKey;
import cd.connect.app.config.DeclaredConfigResolver;
import io.featurehub.web.security.oauth.providers.OAuth2Provider;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Inject;
import javax.ws.rs.client.Client;
import javax.ws.rs.client.Entity;
import javax.ws.rs.core.Form;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

public class OAuth2JerseyClient implements OAuth2Client {
  private static final Logger log = LoggerFactory.getLogger(OAuth2JerseyClient.class);
  protected final Client client;

  // the url we pass to the POST to confirm we are who we say we are
  @ConfigKey("oauth2.redirectUrl")
  protected String redirectUrl;

  @Inject
  public OAuth2JerseyClient(Client client) {
    this.client = client;

    DeclaredConfigResolver.resolve(this);
  }

  public AuthClientResult requestAccess(String code, OAuth2Provider provider) {
    Form form = new Form();
    form.param("grant_type", "authorization_code");
    form.param("client_id", provider.getClientId());
    form.param("client_secret", provider.getClientSecret());
    form.param("redirect_uri", redirectUrl);
    form.param("code", code);

    Entity<Form> entity = Entity.entity(form, MediaType.APPLICATION_FORM_URLENCODED_TYPE);

    final Response response = client.target(provider.requestTokenUrl()).request()
      .accept(MediaType.APPLICATION_JSON)
      .post(entity);
    if (response.getStatusInfo().getFamily() == Response.Status.Family.SUCCESSFUL) {
      return response.readEntity(AuthClientResult.class);
    } else {
      log.warn("OAuth2 Login attempt failed! {}", response.getStatus());
      return null;
    }
  }
}

File: backend/security-oauth/src/main/kotlin/io/featurehub/web/security/oauth/OAuth2JerseyClient.kt
package io.featurehub.web.security.oauth

import cd.connect.app.config.ConfigKey
import cd.connect.app.config.DeclaredConfigResolver
import io.featurehub.web.security.oauth.providers.OAuth2Provider
import jakarta.inject.Inject
import jakarta.ws.rs.client.Client
import jakarta.ws.rs.client.Entity
import jakarta.ws.rs.core.Form
import jakarta.ws.rs.core.MediaType
import jakarta.ws.rs.core.Response
import org.slf4j.LoggerFactory
import java.util.*

class OAuth2JerseyClient @Inject constructor(protected val client: Client) : OAuth2Client {
  // the url we pass to the POST to confirm we are who we say we are
  @ConfigKey("oauth2.redirectUrl")
  protected var redirectUrl: String? = null
  override fun requestAccess(code: String?, provider: OAuth2Provider): AuthClientResult? {
    val form = Form()
    form.param("grant_type", "authorization_code")
    form.param("client_id", provider.clientId)

    if (!provider.isSecretInHeader()) {
      form.param("client_secret", provider.clientSecret)
    }

    form.param("redirect_uri", redirectUrl)
    form.param("code", code)
    val target = client.target(provider.requestTokenUrl())
    var request = target.request()

    provider.enhanceTokenRequest(request, form)
    val entity = Entity.entity(form, MediaType.APPLICATION_FORM_URLENCODED_TYPE)

    if (provider.isSecretInHeader()) {
      val code = "${provider.clientId}:${provider.clientSecret}"
      request = request.header("Authorization",
        "Basic " + Base64.getEncoder().encodeToString(code.toByteArray()))
    }

    val response = request
      .accept(MediaType.APPLICATION_JSON)
      .post(entity)
    return if (response.statusInfo.family == Response.Status.Family.SUCCESSFUL) {
      response.readEntity(AuthClientResult::class.java)
    } else {
      log.warn("OAuth2 Login attempt failed! {}", response.status)
      null
    }
  }

  companion object {
    private val log = LoggerFactory.getLogger(OAuth2JerseyClient::class.java)
  }

  init {
    DeclaredConfigResolver.resolve(this)
  }
}
